INX

alexsporn opened the pull request:
- Added basic INX support
- Re-implemented Indexer and MQTT plugins as INX extensions

alexsporn committed:
- Added empty INX plugin

alexsporn committed:
- Refactored API message attachment (tipsel+pow) into reusable MessageAttacher

alexsporn committed:
- Added initial INX plugin and a inx-client that streams all the new received messages

alexsporn committed:
- Refactored Indexer API into indexer_server pkg

alexsporn committed:
- Added new inx-indexer tool that implements the same indexer plugin but over INX

alexsporn committed:
- Removed external ledger locking calls from INX

- Changed how the indexer syncs over INX

alexsporn committed:
- Add a reverse proxy to expose further API endpoints over INX

alexsporn committed:
- Register inx-indexer API route

- Add signal handling and proper deregister and shutdown

alexsporn committed:
- Start and stop the inx-indexer from within HORNET

alexsporn committed:
- Read node status and decide if indexer needs to reset the local database or if it can be synced

alexsporn committed:
- Use random port for the indexer http server since it is registered over INX anyway

alexsporn committed:
- Adding output.log and err.log to the extensions executions

- Using ioutil.ReadDir instead of filepath.Walk since we do not need to walk the whole tree
- Send signal to stop extension and kill it if the signal returns an error

alexsporn committed:
- Add missing nil check

alexsporn committed:
- Added a simple Javascript INX extension

alexsporn committed:
- Add inx-nodejs node_modules to gitignore

alexsporn committed:
- Added new endpoint to INX to listen to all referenced messages and get their ledger inclusion state, etc.

alexsporn committed:
- Refactored INX proto interface and added much more requests and events

- Refactored all INX Listen functions to use workerpools, so they do not slow down the internal data flow

alexsporn committed:
- Changed the ReadUnspentOutputs call to send the ledgerIndex for each output, so that the indexer can properly set it's ledgerIndex after initial import

alexsporn committed:
- Disable INX plugin by default and enable it over the config.json

alexsporn committed:
- Removed built-in MQTT plugin

- Added new INX MQTT extension
- Refactored MessageAttacher to use options
- Refactored tip score calculation into a new tangle.TipScoreCalculator. Moved the corresponding settings to the tangle pkg.
-

alexsporn committed:
- Avoid capturing loop var in go func

alexsporn committed:
- Added missing error check and fixed unlock

alexsporn committed:
- Added missing error check

alexsporn committed:
- Fixed goimports

alexsporn committed:
- Fixed some race conditions in the MQTT topic subscription handling

alexsporn committed:
- Passing the INX port to the extensions as INX_PORT env variable

alexsporn committed:
- Added config.json to MQTT extension

alexsporn committed:
- Removed indexer dependency from faucet

alexsporn committed:
- Removed indexer plugin and replaced it with the INX version

alexsporn committed:
- ReadNodeStatus only returns all the milestone information if lmi/cmi are newer than the pruning index

alexsporn committed:
- Added inx.json manifest file to the inx extensions instead of running run.sh directly

alexsporn committed:
- Build Indexer and MQTT inside the Dockerfile

alexsporn committed:
- Cleaned up Indexer plugin from private_tangle scripts

alexsporn committed:
- Enable INX on integration tests doing balance calculations

alexsporn committed:
- Fixed building INX extensions inside Dockerfile

alexsporn committed:
- Removed nodejs from inx folder

alexsporn committed:
- Build INX extensions when build HORNET with the scripts or when running the private_tangle coordinator scripts

alexsporn committed:
- Added config to INX plugin

- Added a way to disable specific INX extensions using the config or flags

alexsporn committed:
- Added migration receipt topic handling to MQTT plugin

muXxer commented on .gitignore:
```diff
@@ -45,3 +45,11 @@ coordinator.state
 book
 stardust_testnet
+tools/inx-nodejs/node_modules
+inx/indexer/err.log
```
Why not just a wildcard for all `err.log`, `output.log`, etc?

muXxer commented on config.json:
```diff
@@ -18,6 +17,8 @@
       "/api/plugins/debug/v1/*",
       "/api/plugins/indexer/v1/*",
+      "/api/plugins/mqtt/v1",
+      "/api/plugins/inx-*",
```
```suggestion
```

muXxer replied:
```suggestion
```

muXxer commented on tools/inx-nodejs/client.js:
```diff
@@ -0,0 +1,74 @@
+const util = require("util");
+const PROTO_PATH = "../../pkg/inx/proto/inx.proto";
+
+const options = {
```
Maybe we should rename it to `protoOptions` or something.
People will copy this JS example and I thinks its more readable this way.

muXxer commented on tools/inx-nodejs/client.js:
```diff
@@ -0,0 +1,74 @@
+const Koa = require('koa');
```
Maybe we should add a README.md what this example is doing. :)

muXxer commented on tools/inx-nodejs/client.js:
```diff
@@ -0,0 +1,74 @@
+const app = new Koa();
+const router = new Router();
+
+router.get('/inx-nodejs/v1/info', (ctx) => {
```
Maybe add a comment what this actually does.

Exposing a new route with info from the node via INX.

muXxer commented on tools/inx-nodejs/client.js:
```diff
@@ -0,0 +1,74 @@
+
+const apiRouteRequest = {"route": "inx-nodejs/v1", "host": "localhost", "port": 3000};
+
+client.RegisterAPIRoute(apiRouteRequest, (err) => {
```
Add a comment that this registers a new route in the node itself.

muXxer commented on tools/inx-nodejs/client.js:
```diff
@@ -0,0 +1,74 @@
+
+client.RegisterAPIRoute(apiRouteRequest, (err) => {
+    if (err) {
+        console.log(err);
```
Maybe exit the process if it was not successful?

muXxer commented on private_tangle/run_3rd.sh:
```diff
@@ -6,14 +6,13 @@ go run ../main.go -c config_private_tangle.json \
 --db.path="privatedb3" \
 --node.disablePlugins="Autopeering" \
---node.enablePlugins="Spammer,MQTT,Debug,Prometheus,Indexer" \
+--node.enablePlugins="Spammer,MQTT,Debug,Prometheus" \
```
```suggestion
--node.enablePlugins="Spammer,Debug,Prometheus" \
```

muXxer replied:
```suggestion
--node.enablePlugins="Spammer,Debug,Prometheus" \
```

muXxer commented on private_tangle/run_coo_bootstrap.bat:
```diff
@@ -11,7 +11,7 @@ go run "..\main.go" -c config_private_tangle.json ^
 --db.path="privatedb" ^
 --node.disablePlugins="Autopeering" ^
---node.enablePlugins="Spammer,Coordinator,MQTT,Debug,Prometheus,Faucet,Indexer" ^
+--node.enablePlugins="Spammer,Coordinator,Debug,Prometheus,Faucet,INX" ^
```
remove the `--mqtt.bindAddress="localhost:1883" ^` in this file as well

muXxer replied:
remove the `--mqtt.bindAddress="localhost:1883" ^` in this file as well

muXxer commented on pkg/inx/proto/inx.proto:
```diff
@@ -0,0 +1,170 @@
+syntax = "proto3";
+
+package inx;
+option go_package = "github.com/gohornet/pkg/inx";
```
```suggestion
option go_package = "github.com/gohornet/hornet/pkg/inx";
```

muXxer commented on pkg/model/faucet/faucet.go:
```diff
@@ -325,21 +320,35 @@ func (f *Faucet) Info() (*FaucetInfoResponse, error) {
+		return true
 	}
-	return amount, nil
+	if err := f.utxoManager.ForEachUnspentOutput(consumerFunc, utxo.ReadLockLedger(false)); err != nil {
```
Isn't that quite heavy computational wise?
Can you bridge the request to the indexer with your new interface instead?

alexsporn replied:
I would do that in a later PR

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+}
+
+func (s *Server) onSubscribeTopic(ctx context.Context, topic string) {
+	if topic == topicMilestonesLatest {
```
Why don't you use a switch case here?

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+}
+
+func (s *Server) onUnsubscribeTopic(topic string) {
+	if topic == topicMilestonesLatest {
```
Why don't you use a switch case here?

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+		} else {
+			sub.Count--
+		}
+	}
```
```suggestion
	if ok {
		sub.Count--
		if sub.Count == 0 {
			sub.Func()
			delete(s.grpcSubscriptions, grpcCall)
		}
	}
```

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+		subscriptionIdentifier := rand.Int()
+		s.grpcSubscriptions[grpcCall] = &topicSubcription{
+			Count:      1,
+			Func:       cancel,
```
Maybe rename to `CancelFunc`.... it wasn't clear to me when I saw the `Func()` in `stopListenIfNeeded`

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			s.grpcSubscriptionsLock.Unlock()
+		}()
+	} else {
+		sub.Count++
```
I would swap the logic to `if ok {`, increase the count do an early return.
```go
if ok {
	sub.Count++
	return
}

```

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			break
+		}
+		if c.Err() != nil {
+			break
```
why don't we print the same log here like in the above error case?

alexsporn replied:
Because it will always print "context cancelled"

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			break
+		}
+		if c.Err() != nil {
+			break
```
why don't we print the same log here like in the above error case?

alexsporn replied:
Because it will always print "context cancelled"

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			break
+		}
+		if c.Err() != nil {
+			break
```
why don't we print the same log here like in the above error case?

alexsporn replied:
Because it will always print "context cancelled"

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			break
+		}
+		if c.Err() != nil {
+			break
```
why don't we print the same log here like in the above error case?

alexsporn replied:
Because it will always print "context cancelled"

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			break
+		}
+		if c.Err() != nil {
+			break
```
why don't we print the same log here like in the above error case?

alexsporn replied:
Because it will always print "context cancelled"

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			break
+		}
+		if c.Err() != nil {
+			break
```
why don't we print the same log here like in the above error case?

alexsporn replied:
Because it will always print "context cancelled"

muXxer commented on tools/inx-mqtt/server.go:
```diff
@@ -0,0 +1,408 @@
+			break
+		}
+		if c.Err() != nil {
+			break
```
why don't we print the same log here like in the above error case?

alexsporn replied:
Because it will always print "context cancelled"

muXxer commented on tools/inx-client/main.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+const (
+	INXPort = 9029
```
Maybe add the env var code in this example as well.

muXxer commented on pkg/tangle/attacher.go:
```diff
@@ -0,0 +1,148 @@
+		msg.Parents = tips.ToSliceOfArrays()
+	}
+
+	if msg.Nonce == 0 {
```
Shouldn't we always check if the score is sufficient?

alexsporn replied:
It will be checked inside Emit() again. This is only done in the `Nonce == 0` case because 0 might be a valid nonce and we don't want to do PoW for those cases.

muXxer commented on plugins/inx/extension.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func NewExtension(path string) (*Extension, error) {
+	//TODO: read inx.json or some config file
```
```suggestion
```

muXxer commented on plugins/inx/extension.go:
```diff
@@ -0,0 +1,86 @@
+		return nil, err
+	}
+	if len(config.String(INXManifestName)) == 0 {
+		return nil, fmt.Errorf("inx.json missing key %s", INXManifestName)
```
Include the path in the error message, so we know which extension is missing the parameter.

muXxer commented on plugins/inx/extension.go:
```diff
@@ -0,0 +1,86 @@
+		return nil, fmt.Errorf("inx.json missing key %s", INXManifestName)
+	}
+	if len(config.String(INXManifestEntrypoint)) == 0 {
+		return nil, fmt.Errorf("inx.json missing key %s", INXManifestEntrypoint)
```
Include the path in the error message, so we know which extension is missing the parameter.

muXxer commented on plugins/inx/extension.go:
```diff
@@ -0,0 +1,86 @@
+	e.cmd.Dir = e.Path
+
+	var logFile *os.File
+	logFile, err := os.Create(filepath.Join(e.Path, "output.log"))
```
Add inx config parameters to disable the logging of the forked process.

alexsporn committed:
- Added a dynamic proxy to add/remove reverse proxies to INX plugins

alexsporn committed:
- Added INX (gRPC) metrics to prometheus

alexsporn committed:
- Added configuration file to Indexer plugin

- Added Prometheus metrics to Indexer plugin

alexsporn committed:
- Added Prometheus metrics to MQTT plugin

alexsporn committed:
- Exposing INX plugins Prometheus metrics over the same HTTP server the node metrics are exposed with

alexsporn committed:
- Added MQTT topics_manager_size metric to prometheus

muXxer commented on tools/inx-indexer/main.go:
```diff
@@ -0,0 +1,278 @@
+	messageID := output.UnwrapMessageID()
+	milestoneIndex := milestone.Index(output.GetMilestoneIndexBooked())
+	milestoneTimestamp := uint64(output.GetMilestoneTimestampBooked())
+	o, err := output.UnwrapOutput(serializer.DeSeriModePerformValidation)
```
You wanted to perform validation here?

muXxer commented on tools/inx-indexer/main.go:
```diff
@@ -0,0 +1,278 @@
+		if err := importer.AddOutput(output); err != nil {
+			return err
+		}
+		if outputLedgerIndex > ledgerIndex {
```
I like when both is on the same side after the condition :innocent: 
```suggestion
		if ledgerIndex < outputLedgerIndex {
```

muXxer commented on tools/inx-indexer/main.go:
```diff
@@ -0,0 +1,278 @@
+	e.Use(middleware.Recover())
+
+	go func() {
+		if err := e.Start(":0"); err != nil {
```
what is that doing here exactly? Can you add a comment? :innocent:

alexsporn replied:
This starts the server with a random (free) port. I already changed it to use the bindAddress from the config

muXxer commented on pkg/restapi/proxy.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func (b *balancer) skipper(c echo.Context) bool {
+	b.mutex.RLock()
```
why do you lock here? the `Next` already locks internally?

muXxer commented on plugins/restapi/v2/messages.go:
```diff
@@ -78,30 +75,27 @@ func messageMetadataByID(c echo.Context) (*messageMetadataResponse, error) {
-			shouldReattach = false
-		} else if (cmi - ocri) > milestone.Index(deps.MaxDeltaMsgOldestConeRootIndexToCMI) {
-			// if the OCRI to CMI delta is over CfgTipSelMaxDeltaMsgOldestConeRootIndexToCMI, the tip is semi-lazy and should be promoted
+		switch tipScore {
```
Would it make sense to include the healthy case without further instructions, for better readability?

muXxer commented on plugins/restapi/plugin_manager.go:
```diff
@@ -0,0 +1,79 @@
+	if !found {
+		p.plugins = append(p.plugins, pluginRoute)
+	}
+	return p.proxy.AddGroup(pluginRoute)
```
We overwrite the old one in case it was found? Or what would happen here?

alexsporn replied:
Yes, that is important if the extension crashed/shutdown for some reason and subscribes again

muXxer replied:
```suggestion
	// existing groups get overwritten (necessary if last plugin was not cleaned up properly)
	return p.proxy.AddGroup(pluginRoute)
```

muXxer commented on plugins/restapi/plugin_manager.go:
```diff
@@ -0,0 +1,79 @@
+	if !found {
+		p.plugins = append(p.plugins, pluginRoute)
+	}
+	p.proxy.AddReverseProxy(pluginRoute, host, port)
```
We overwrite the old one in case it was found? Or what would happen here?

alexsporn replied:
Yes, that is important if the extension crashed/shutdown for some reason and subscribes again

muXxer replied:
```suggestion
	// existing proxies get overwritten (necessary if last plugin was not cleaned up properly)
	p.proxy.AddReverseProxy(pluginRoute, host, port)
```

alexsporn committed:
- Fixed review comments

alexsporn committed:
- Added maxPageSize config to Indexer

- Read the bech32 prefix from INX instead of hardcoding it

alexsporn committed:
- Refactored rest API error handler into restapi pkg and used it both the v2 api and in the indexer

alexsporn committed:
- Added a new PerformAPIRequest grpc call to INX to allow plugins to perform HTTP API requests

- Added a custom http.RoundTripper implementation that uses PerformAPIRequest. This allows us to use the default iota.go nodeclient to query e.g. the Indexer plugin

alexsporn committed:
- Updated iota.go/v3

alexsporn committed:
- Adapted tests to latest foundry changes

alexsporn committed:
- Generated new snapshot files for the integration tests

alexsporn committed:
- Disable gzip middleware in Indexer, since else we will have double gzip compression. The normal rest api server is already compressing the responses

alexsporn committed:
- Using iotago.ZeroRentParas for some tests and setting default VByteCost to 500

alexsporn committed:
- Check for Indexer error

muXxer commented on tools/inx-mqtt/metrics.go:
```diff
@@ -0,0 +1,53 @@
+
+	prometheus.MustRegister(mqttBrokerTopicsManagerSize)
+
+	e := echo.New()
```
Should we maybe put this whole stuff of echo init and starting of the server in an util func in hornet?

alexsporn replied:
It is just 3 lines that are always the same, not sure if making an extra lib function out of it makes it more maintainable

muXxer commented on plugins/inx/plugin.go:
```diff
@@ -0,0 +1,172 @@
+		deps.INXServer.Start()
+		startExtensions()
+		<-ctx.Done()
+		stopExtensions()
```
```suggestion
		Plugin.LogInfo("Stopping INX ...")
		stopExtensions()
```

muXxer commented on plugins/inx/plugin.go:
```diff
@@ -0,0 +1,172 @@
+			if err != nil {
+				Plugin.LogErrorf("INX extension ended with error: %s", err)
+			}
+		}()
```
```suggestion
		go func(ext *Extension) {
			Plugin.LogInfof("Starting INX extension: %s", ext.Name)
			err := ext.Start(deps.NodeConfig.Int(CfgINXPort))
			if err != nil {
				Plugin.LogErrorf("INX extension ended with error: %s", err)
			}
		}(e)
```

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+
+		m.ShouldPromote = shouldPromote
+		m.ShouldReattach = shouldReattach
+	}
```
I'm wondering if there could be non-solid milestones.
You wouldn't handle that case yet. :thinking:

alexsporn replied:
This is exactly the same as in the API, so I would ignore it

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+		case tangle.TipScoreOCRIThresholdReached, tangle.TipScoreYCRIThresholdReached:
+			shouldPromote = true
+			shouldReattach = false
+		case tangle.TipScoreBelowMaxDepth:
```
Please also add the TipScoreHealthy case here like in the other file.

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+func (s *INXServer) ReadMessage(_ context.Context, messageID *inx.MessageId) (*inx.RawMessage, error) {
+	cachedMsg := deps.Storage.CachedMessageOrNil(messageID.Unwrap())
```
```suggestion
	cachedMsg := deps.Storage.CachedMessageOrNil(messageID.Unwrap()) // message +1
```

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+	if cachedMsg == nil {
+		return nil, status.Errorf(codes.NotFound, "message %s not found", messageID.Unwrap().ToHex())
+	}
+	defer cachedMsg.Release(true)
```
```suggestion
	defer cachedMsg.Release(true) // message -1
```

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+func (s *INXServer) ReadMessageMetadata(_ context.Context, messageID *inx.MessageId) (*inx.MessageMetadata, error) {
+	cachedMsgMeta := deps.Storage.CachedMessageMetadataOrNil(messageID.Unwrap())
```
```suggestion
	cachedMsgMeta := deps.Storage.CachedMessageMetadataOrNil(messageID.Unwrap()) // meta +1
```

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+	if cachedMsgMeta == nil {
+		return nil, status.Errorf(codes.NotFound, "message metadata %s not found", messageID.Unwrap().ToHex())
+	}
+	defer cachedMsgMeta.Release(true)
```
```suggestion
	defer cachedMsgMeta.Release(true) // meta -1
```

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	wp := workerpool.New(func(task workerpool.Task) {
+		cachedMsg := task.Param(0).(*storage.CachedMessage)
+		defer cachedMsg.Release(true)
```
```suggestion
		defer cachedMsg.Release(true) // message -1
```

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	wp := workerpool.New(func(task workerpool.Task) {
+		msgMeta := task.Param(0).(*storage.CachedMetadata)
+		defer msgMeta.Release(true)
```
```suggestion
		defer msgMeta.Release(true) // meta -1
```

muXxer commented on plugins/inx/server_messages.go:
```diff
@@ -0,0 +1,196 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	wp := workerpool.New(func(task workerpool.Task) {
+		msgMeta := task.Param(0).(*storage.CachedMetadata)
+		defer msgMeta.Release(true)
```
```suggestion
		defer msgMeta.Release(true) // meta -1
```

muXxer commented on plugins/inx/server_milestones.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+func milestoneForIndex(msIndex milestone.Index) (*inx.Milestone, error) {
+	cachedMilestone := deps.Storage.CachedMilestoneOrNil(msIndex)
```
```suggestion
	cachedMilestone := deps.Storage.CachedMilestoneOrNil(msIndex) // milestone +1
```

muXxer commented on plugins/inx/server_milestones.go:
```diff
@@ -0,0 +1,73 @@
+	if cachedMilestone == nil {
+		return nil, status.Errorf(codes.NotFound, "milestone %d not found", msIndex)
+	}
+	defer cachedMilestone.Release(true)
```
```suggestion
	defer cachedMilestone.Release(true) // milestone -1
```

muXxer commented on plugins/inx/server_milestones.go:
```diff
@@ -0,0 +1,73 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	wp := workerpool.New(func(task workerpool.Task) {
+		cachedMilestone := task.Param(0).(*storage.CachedMilestone)
+		defer cachedMilestone.Release(true)
```
```suggestion
		defer cachedMilestone.Release(true) // milestone -1
```

muXxer commented on plugins/inx/server_milestones.go:
```diff
@@ -0,0 +1,73 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	wp := workerpool.New(func(task workerpool.Task) {
+		cachedMilestone := task.Param(0).(*storage.CachedMilestone)
+		defer cachedMilestone.Release(true)
```
```suggestion
		defer cachedMilestone.Release(true) // milestone -1
```

muXxer commented on plugins/inx/server_utxo.go:
```diff
@@ -0,0 +1,185 @@
+			return status.Errorf(codes.InvalidArgument, "given startMilestoneIndex %d is older than the current pruningIndex %d", startIndex, pruningIndex)
+		}
+
+		deps.UTXOManager.ReadLockLedger()
```
Pruning could interfere here, since you lock the ledger after checking the pruning index.
Do we have a pruning lock or something? Or is it sufficient to lock the ledger before?

alexsporn replied:
I am now locking the ledger before, but I think that manual pruning calls are not locking the UTXO ledger.

muXxer commented on plugins/inx/server_utxo.go:
```diff
@@ -0,0 +1,185 @@
+			return status.Errorf(codes.InvalidArgument, "given startMilestoneIndex %d is older than the current pruningIndex %d", startIndex, pruningIndex)
+		}
+
+		deps.UTXOManager.ReadLockLedger()
```
You could put this whole code in the if clause into an inline function, and then do a `defer deps.UTXOManager.ReadUnlockLedger()` here.

muXxer commented on plugins/inx/server_utxo.go:
```diff
@@ -0,0 +1,185 @@
+			return status.Error(codes.Unavailable, "error accessing the UTXO ledger")
+		}
+		currentIndex := startIndex
+		for currentIndex <= ledgerIndex {
```
Why do you increment currentIndex at the end, and not in the for loop statement itself?

alexsporn committed:
- Fixes for review comments

alexsporn committed:
- Added a bool config entry to enable/disable extension log files

alexsporn committed:
- Read port from env in inx-client example

alexsporn committed:
- Document a little bit the nodejs INX example

alexsporn committed:
- go mod tidy after rebase

muXxer commented on plugins/inx/params.go:
```diff
@@ -7,13 +7,13 @@ import (
-
-	// the path with the extensions to be loaded
+	// CfxINXPath the path with the extensions to be loaded
 	CfxINXPath = "inx.path"
```
```suggestion
	CfgINXPath = "inx.path"
```

alexsporn committed:
- Control extension output logging using the extension manifest instead of the global config

alexsporn committed:
- Fixed typo in config constant

alexsporn committed:
- Use --inx.path for the private_tangle instead of the inx folder symlink

alexsporn committed:
- Adapted test amounts to avoid rent cost issues
